base-containers/base/inginious_container_api/utils.py

```diff
@@ -96,3 +105,213 @@ def ssh_wait(ssh_user, timeout=None):
+               "container_id": container_id}
+        event_loop.run_until_complete(write_stdout(msg, container_stdout))
+    # Wait for user to connect and leave
+    ssh_retval = ssh_wait(ssh_user)
+    return ssh_retval
+
+
+def receive_initial_command(both_dockers, container_stdin, event_loop):
+    """ Receive the command to run (directly from student-grading socket if both dockers or via the agent otherwise)"""
+    if both_dockers:  # Grading and student containers are both on docker
+        # Connect to the socket
+        my_socket = socket.socket(socket.AF_UNIX)  # , socket.SOCK_CLOEXEC) # for linux only
+        my_socket.connect("/__parent.sock")
+        # Say hello
+        print("Saying hello")
+        my_socket.send(b'H')
+        print("Said hello")
+        # Receive fds
+        print("Receiving fds")
+        msg, fds = recv_fds(my_socket, 1, 3)
+        assert msg == b'S'
+        print("Received fds")
+        # Unpack the start message
+        print("Unpacking start cmd")
+        unpacker = msgpack.Unpacker()
+        start_cmd = None
+        while start_cmd is None:
+            data = my_socket.recv(1)
+            unpacker.feed(data)
+            for obj in unpacker:
+                return my_socket, fds, obj
+    else:  # Grading or student container is on Kata
+        msg = event_loop.run_until_complete(receive_initial_message(container_stdin))
+        return None, None, msg
+
+
+async def receive_initial_message(reader: asyncio.StreamReader):
+    """ Get the initial command message from the agent.
+     Used only when both containers are not on a shared-kernel """
+    buf = bytearray()
+    while len(buf) != 4 and not reader.at_eof():
+        buf += await reader.read(4 - len(buf))
+    length = struct.unpack('!I', bytes(buf))[0]
+    buf = bytearray()
+    while len(buf) != length and not reader.at_eof():
+        buf += await reader.read(length - len(buf))
+    message = msgpack.unpackb(bytes(buf), use_list=False)  # EXCEPTION EXTRA DATA
+    return message
+
+
+async def stdio():
+    """ Create the stdin and stdout streams to communicate with the agent """
+    my_loop = asyncio.get_event_loop()
+    reader = asyncio.StreamReader()
+    reader_protocol = asyncio.StreamReaderProtocol(reader)
+    writer_transport, writer_protocol = await my_loop.connect_write_pipe(asyncio.streams.FlowControlMixin, os.fdopen(1, 'wb'))
+    writer = asyncio.StreamWriter(writer_transport, writer_protocol, None, my_loop)
+    await my_loop.connect_read_pipe(lambda: reader_protocol, sys.stdin)
+    return reader, writer
+
+
+def handle_stdin_message(msg, proc_input, proc):
+    """ Process a single message from the agent (stdin message for the student code process or signals messages).
+    Used only when both containers are not on a shared kernel """
+    try:
+        if msg["type"] == "stdin":
+            input_content = msg["message"] + '\n'
+            proc_input.write(input_content.encode("utf8"))
```

Comment:
Again, only text supported here I guess

Reply:
Same comment as previous one, now all the streams are binaries.